File: backend/firebase_config.py
import firebase_admin
from firebase_admin import credentials, firestore, auth
import os
import json
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

def init_firebase():
    """Initialize Firebase Admin SDK"""
    try:
        # Check if already initialized
        firebase_admin.get_app()
    except ValueError:
        # Create credentials from environment variables
        firebase_config = {
            "type": os.getenv("FIREBASE_TYPE"),
            "project_id": os.getenv("FIREBASE_PROJECT_ID"),
            "private_key_id": os.getenv("FIREBASE_PRIVATE_KEY_ID"),
            "private_key": os.getenv("FIREBASE_PRIVATE_KEY").replace('\\n', '\n'),
            "client_email": os.getenv("FIREBASE_CLIENT_EMAIL"),
            "client_id": os.getenv("FIREBASE_CLIENT_ID"),
            "auth_uri": os.getenv("FIREBASE_AUTH_URI"),
            "token_uri": os.getenv("FIREBASE_TOKEN_URI"),
            "auth_provider_x509_cert_url": os.getenv("FIREBASE_AUTH_PROVIDER_X509_CERT_URL"),
            "client_x509_cert_url": os.getenv("FIREBASE_CLIENT_X509_CERT_URL")
        }
        
        # Initialize Firebase with service account credentials
        cred = credentials.Certificate(firebase_config)
        firebase_admin.initialize_app(cred)
    
    return firestore.client()

def get_db():
    """Get Firestore database instance"""
    return firestore.client()

def verify_firebase_token(token):
    """Verify Firebase ID token"""
    try:
        decoded_token = auth.verify_id_token(token)
        return decoded_token
    except Exception as e:
        print(f"Token verification failed: {e}")
        return None

def create_custom_token(uid):
    """Create custom token for user"""
    try:
        custom_token = auth.create_custom_token(uid)
        return custom_token
    except Exception as e:
        print(f"Token creation failed: {e}")
        return None

File: backend/firebaseConfig.js
import admin from 'firebase-admin';
import dotenv from 'dotenv';

// Load environment variables
dotenv.config();

let db = null;

export function initFirebase() {
    try {
        // Check if already initialized
        admin.app();
    } catch (error) {
        // Create credentials from environment variables
        const firebaseConfig = {
            type: process.env.FIREBASE_TYPE,
            project_id: process.env.FIREBASE_PROJECT_ID,
            private_key_id: process.env.FIREBASE_PRIVATE_KEY_ID,
            private_key: process.env.FIREBASE_PRIVATE_KEY?.replace(/\\n/g, '\n'),
            client_email: process.env.FIREBASE_CLIENT_EMAIL,
            client_id: process.env.FIREBASE_CLIENT_ID,
            auth_uri: process.env.FIREBASE_AUTH_URI,
            token_uri: process.env.FIREBASE_TOKEN_URI,
            auth_provider_x509_cert_url: process.env.FIREBASE_AUTH_PROVIDER_X509_CERT_URL,
            client_x509_cert_url: process.env.FIREBASE_CLIENT_X509_CERT_URL
        };
        
        // Initialize Firebase with service account credentials
        const credential = admin.credential.cert(firebaseConfig);
        admin.initializeApp({ credential });
    }
    
    db = admin.firestore();
    return db;
}

export function getDb() {
    if (!db) {
        db = admin.firestore();
    }
    return db;
}

export async function verifyFirebaseToken(token) {
    try {
        const decodedToken = await admin.auth().verifyIdToken(token);
        return decodedToken;
    } catch (error) {
        console.log(`Token verification failed: ${error}`);
        return null;
    }
}

export async function createCustomToken(uid) {
    try {
        const customToken = await admin.auth().createCustomToken(uid);
        return customToken;
    } catch (error) {
        console.log(`Token creation failed: ${error}`);
        return null;
    }
}